packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -0,0 +1,215 @@
+      filteredNode.push(node);
+    }
+  }
+  return filteredNode;
+}
+
+export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>): ComboBoxState<T> {
+  let {
+    onFilter,
+    collator,
+    onSelectionChange
+  } = props;
+
+  let [isFocused, setFocused] = useState(false);
+  let itemsControlled = !!onFilter;
+
+  let computeKeyFromValue = (value, collection) => {
+    let key;
+    for (let [itemKey, node] of collection.keyMap) {
+      if (node.type !== 'section') {
+        let itemText = node.textValue;
+        if (itemText === value) {
+          key = itemKey;
+          break;
+        }
+      }
+    }
+
+    return key;
+  };
+
+  // Need this collection here so that an initial inputValue can be found via collection.getItem
+  // This is really just a replacement for using CollectionBuilder
+  let {collection} = useSingleSelectListState({
+    ...props,
+    // default to null if props.selectedKey isn't set to avoid useControlledState's uncontrolled to controlled warning
+    selectedKey: props.selectedKey || null
+  });
+
+  if (props.selectedKey && props.inputValue) {
+    let selectedItem = collection.getItem(props.selectedKey);
+    let itemText = selectedItem ? selectedItem.textValue : '';
+    if (itemText !== props.inputValue) {
+      throw new Error('Mismatch between selected item and inputValue!');
+    }
+  }
+
+  let onInputChange = (value) => {
+    if (props.onInputChange) {
+      props.onInputChange(value);
+    }
+
+    let newSelectedKey = computeKeyFromValue(value, collection);
+    if (newSelectedKey !== selectedKey) {
+      if (onSelectionChange) {
+        onSelectionChange(newSelectedKey);
+      }
+    }
+  };
+
+  let initialSelectedKeyText = collection.getItem(props.selectedKey)?.textValue;
+  let initialDefaultSelectedKeyText = collection.getItem(props.defaultSelectedKey)?.textValue;
+  let [inputValue, setInputValue] = useControlledState(toString(props.inputValue), initialSelectedKeyText || toString(props.defaultInputValue) || initialDefaultSelectedKeyText || '', onInputChange);
+
+  let selectedKey = props.selectedKey || computeKeyFromValue(inputValue, collection);
+
+  let triggerState = useMenuTriggerState(props);
+
+  // Fires on selection change (when user hits Enter, clicks list item, props.selectedKey is changed)
+  let setSelectedKey = useCallback((key) => {
+    let item = collection.getItem(key);
+    let itemText = item ? item.textValue : '';
+    // think about the below conditionals below
+    // If I don't have the extra itemText check, then setting props.selectedKey to undef or just deleting one letter of the text
+    // so it doesn't match a key will then clear the textfield entirely (in controlled selected key case)
+    // Problem with this is that a clear button w/ setSelectedKey = '' won't actually clear the textfield because of this itemText check (also controlled selected key only case, input value uncontrolled).
+    itemText && setInputValue(itemText);
```

Comment:
Try the "Clear key" button in any of the `selectedKey` stories in the story book. Note that it doesn't clear the text field because of this `itemText &&` check. The predicament that arises is that removing the `itemText &&` check from here will make the ComboBox clear the entire input once you type in a value that doesn't match any of the ComboBox values for those same `selectedKey` stories.

E.g. for the ControlledKey story
1. Type "Kangaroo" (selectedKey becomes 2)
2. Delete the last 'o' from the field
3. This triggers `onInputChange` -> `onSelectionChange(null)` since the input value no longer matches a valid combobox value
4. `onSelectionChange` fires `setSelectedKey(null)` (setter from the controlled key story) which updates `prop.selectedKey` to null
5. The update to `prop.selectedKey` fires the `useEffect` in line 132 which calls `setSelectedKey(null)` (the function on line 111)
6. `setInputText` gets called with `''`since `itemText = ''` since the key was null, clearing the whole field instead of leaving the field with "Kangaro"